### PR TITLE
:sparkles: SAT-based clock number assignment

### PR DESCRIPTION
**Pull request by marcelwa**
## Description

This PR introduces a new SAT-based algorithm that can determine valid clock number assignments to gate-level layouts. It ignores the present clocking information of the given layout and only looks at the gate-wire connections present in the layout. Based on these connections, it creates a SAT instance that formulates the clock number assignment problem. This algorithm is exact; that is, if a valid clock number assignment exists, it will eventually be found. Vice versa, if no clock number assignment can be found, none exists. In the former case, the algorithm will override the clocking information present in the layout via the assignment of an irregular clock map.

## Checklist:

- [x] The pull request only contains commits that are related to it.
- [x] I have added appropriate tests and documentation.
- [ ] I have made sure that all CI jobs on GitHub pass.
- [x] The pull request introduces no new warnings and follows the project's style guidelines.

### CODE REVIEW

File: docs/algorithms/determine_clocking.rst [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+Determine Clocking
+------------------
+
+**Header:** ``fiction/algorithms/physical_design/determine_clocking.hpp``
+
+Determines a new clock number assignment for the given gate-level layout. Whether the layout has useful/valid clocking
+information or not, this function will override each gate and wire segment with new clocking information if a valid
+clock number assignment exists.
+
+.. doxygenfunction:: fiction::determine_clocking
```
Comment (Drewniok) [inline docs/algorithms/determine_clocking.rst:12]: ``params`` is missing.

File: include/fiction/algorithms/physical_design/determine_clocking.hpp [hunk 2]
```diff
@@ -0,0 +1,425 @@
...
+ * If no valid clock number assignment exists for `lyt`, this function returns `false` and does not modify `lyt`.
+ *
+ * @tparam Lyt Gate-level layout type.
+ * @param lyt The gate-level layout to assign clock numbers to.
+ * @param params Parameters.
+ * @param stats Statistics.
+ * @return `true` iff `lyt` could be successfully clocked via a valid clock number assignment.
+ */
+template <typename Lyt>
+bool determine_clocking(Lyt& lyt, assign_clocking_params params = {}, assign_clocking_stats* stats = nullptr)
```
Comment (Drewniok) [inline include/fiction/algorithms/physical_design/determine_clocking.hpp:406]: Parameters can be marked with ``const``, right?

**Review by simon1hofmann — DISMISSED** [review]
Thanks a lot for the addition @marcelwa.
Looks really good 👍

**Review by Drewniok — APPROVED** [review]
LGTM, thanks for implementing this!

**Pull request merged**